lib/chef/http.rb

```diff
@@ -329,6 +329,13 @@ def retrying_http_errors(url)
           retry
         end
         raise Timeout::Error, "Timeout connecting to #{url}, giving up"
+      rescue OpenSSL::SSL::SSLError
+        if http_retry_count - http_attempts + 1 > 0
+          Chef::Log.error("SSL Error connecting to #{url}, retry #{http_attempts}/#{http_retry_count}")
+          sleep(http_retry_delay)
+          retry
+        end
+        raise Timeout::Error, "SSL Error connecting to #{url}, giving up"
```

Comment:
Changing the error class here will mess with some stuff in knife that gives you hints about what to do next when you get an SSL error. The most common cause of this error is that you set up a Chef Server using the default self-generated, self-signed certificate, which cannot be verified with any cert in the default CA bundle. Knife rescues `SSLError` and tells you that you can possibly fix it by running `knife ssl check` and `knife ssl fetch`. I'd also like to add that kind of check and error messaging to chef-client. I think it would also be a good idea to skip the retries for certificate verification failures, since we can be confident that a retry will not fix the issue, so there is no point in forcing the user to wait for the retries to all fail.

It would probably be fine to change the error class to something under `Chef::Exceptions` though, which would maybe give us the ability to store enough information in the exception to fetch the server's certificate and then run a more useful SSL verification function against it (see https://gist.github.com/danielsdeleo/1ddecfaf638a8b23b8bb for a basic idea of the implementation). If we do that, then we should also update knife to accommodate the new error type. That code is here: https://github.com/chef/chef/blob/master/lib/chef/knife.rb#L439-L445